src/inmanta/server/services/environment_metrics_service.py

```diff
@@ -0,0 +1,120 @@
+        should be interpreted as [start_interval, end_interval[
+
+        :param start_interval: The start time of the metrics collection interval (inclusive).
+        :param end_interval: The end time of the metrics collection interval (exclusive).
+        :result: The metrics collected by this MetricCollector within the past metrics collection interval.
+        """
+        raise NotImplementedError()
+
+
+class EnvironmentMetricsService(protocol.ServerSlice):
+    """Slice for the management of metrics"""
+
+    def __init__(self) -> None:
+        super(EnvironmentMetricsService, self).__init__(SLICE_ENVIRONMENT_METRICS)
+        self.metrics_collectors: List[MetricsCollector] = []
+
+    def get_dependencies(self) -> List[str]:
+        return [SLICE_DATABASE]
+
+    def get_depended_by(self) -> List[str]:
+        return [SLICE_TRANSPORT]
+
+    async def start(self) -> None:
+        await super().start()
+        self.schedule(self.flush_metrics, 60, initial_delay=0, cancel_on_stop=True)
+
+    def register_metric_collector(self, metrics_collector: MetricsCollector) -> None:
+        """
+        Register the given metrics_collector.
+        """
+        if not any(metric.get_metric_name() == metrics_collector.get_metric_name() for metric in self.metrics_collectors):
+            self.metrics_collectors.append(metrics_collector)
+        else:
+            LOGGER.warning(f"There already is a metric collector with the name {metrics_collector.get_metric_name()}")
+
+    async def flush_metrics(self) -> None:
+        """
+        Invoked at the end of the metrics collection interval. Writes the metrics
+        collected by the MetricsCollectors in the past metrics collection interval
+        to the database.
+        """
+        now: datetime = datetime.now()
+        metric_count: List[EnvironmentMetricsCounter] = []
+        metric_non_count: List[EnvironmentMetricsNonCounter] = []
+        for metrics_collector in self.metrics_collectors:
+            metric_name: str = metrics_collector.get_metric_name()
+            metric_type: str = metrics_collector.get_metric_type()
+            metric_value: dict[str, int] = await metrics_collector.get_metric_value(now - timedelta(seconds=60), now)
+            if metric_type == MetricType.count:
+                metric_count.append(EnvironmentMetricsCounter.new(metric_name, now, metric_value["count"]))
+            if metric_type == MetricType.non_count:
+                metric_non_count.append(
+                    EnvironmentMetricsNonCounter.new(metric_name, now, metric_value["count"], metric_value["value"])
+                )
+
+        await EnvironmentMetricsCounter.insert_many(metric_count)
+        await EnvironmentMetricsNonCounter.insert_many(metric_non_count)
+
+        if datetime.now() - now > timedelta(seconds=60):
+            LOGGER.warning("flush_metrics method took more than 1 minute")
```

Comment:
I would hard code the 60 second here, but use the constant that defines the collection interval.